content/post/graceful-k8s-delpoyments/index.md

```diff
@@ -0,0 +1,205 @@
+---
+title: "Zero-Downtime Rolling Deployments in Kubernetes"
+date: 2021-07-1500:00:00+09:00
+hero: /images/rolling-waves.jpg
+excerpt: |
+  Kubernetes enables us to operate robust systems at scale. However, it is not perfect and in edge cases k8s may behave in unexpected ways.
+authors:
+  - Manish Yadav
+  - Oscar Garcia
+  - Nemo
+  - Boaz Yaniv
+draft: true
+license: cc-by
+---
+
+*[Photo by Kammeran Gonzalez-Keola](https://www.pexels.com/@kammeran-gonzalez-keola-3137381) used under [Pexels License](https://www.pexels.com/license/)*
+
+At Rakuten we default to Kubernetes to run our services, and k8s enables us to operate robust systems at scale. However, it is not perfect and in edge cases k8s may behave in unexpected ways. For example, during rolling deployments we expect no request failures - a reasonable expectation, right? But when we ran functional tests during a deployment we notice requests failing - to our surprise. After researching this for a while we learned that...
+
+> K8s will send traffic to terminating pods.
+
+Even after starting the pod shutdown process, which includes sending a `TERM` signal to pod, k8s still sends new requests to the pod. Because there is no orchestration between k8s sending `TERM` signal and removing the pod from its service endpoint list. Those 2 operations can happen in any order, possibly with delay in between them. Yikes 😬.
+
+In this article, we explore the Kubernetes pod eviction mechanism and show different approaches to avoid these surprising errors in apps deployed with k8s.
+
+## Kubernetes Pod Termination Process
+
+A pod can be evicted for multiple reasons. For example, when a node is drained, via `kubectl delete`, or when the scheduler evicts pods to allow the execution of higher-priority ones. The eviction process starts when the API server modifies the state of a pod in `etcd` to the `Terminating` state. The node’s Kubelet and the endpoints-controller are continuously watching the pod's state. Once they notice the termination state they start the eviction process:
+
+1. Kubelet performs the pod eviction.
+1. The endpoints-controller handles the endpoint removal process.
+
+Both operations are asynchronous.
+
+![SIGTERM in Kubernetes](./images/sigterm.png)
```

Comment:
Please add `Kubelet`, `kube` and `IPtables` to your dictionary 😂

Reply:
not my diagram 😂
originates from oscar's post

Reply:
k, will do

Reply:
I re-created the diagram, without squiggly red lines and every word in lower case (except for all caps SIGTERM and API). lmk if you spot any other problems with the diagram.